amethyst_ui/src/selection_order_cache.rs

```diff
@@ -0,0 +1,128 @@
+use amethyst_core::specs::Entities;
+use amethyst_core::specs::Entity;
+use amethyst_core::specs::Join;
+use amethyst_core::specs::System;
+use amethyst_core::specs::Write;
+use amethyst_core::specs::WriteStorage;
+use hibitset::BitSet;
+use std::cmp::Ordering;
+use std::marker::PhantomData;
+
+use amethyst_core::specs::ReadStorage;
+
+use {Selectable, Selected};
+
+/// A cache sorted by tab order and then by Entity.
+/// Used to quickly find the next or previous selectable entities.
+#[derive(Debug, Clone, Default)]
+pub struct CachedSelectionOrder {
+    /// The cached bitset.
+    pub cached: BitSet,
+    /// The cache holding the z position and the corresponding entity.
+    pub cache: Vec<(u32, Entity)>,
+}
+
+impl CachedSelectionOrder {
+    // TODO: Change WriteStorage to ReadStorage. @torkleyy
+    //
+    //
+    //
+    //
+    //
+    /// Returns the index of the highest cached element (index in the cache!) that is currently selected.
+    pub fn highest_order_selected_index(
+        &self,
+        selected_storage: &WriteStorage<Selected>,
+    ) -> Option<usize> {
+        self.cache
+            .iter()
+            .enumerate()
+            .rev()
+            .find(|(_, (_, e))| selected_storage.get(*e).is_some())
+            .map(|t| t.0)
+    }
+
+    /// Returns the index in the cache for the specified entity.
+    pub fn index_of(&self, entity: Entity) -> Option<usize> {
+        self.cache
+            .iter()
+            .enumerate()
+            .find(|(_, (_, e))| *e == entity)
+            .map(|t| t.0)
+    }
+}
+
+/// System in charge of updating the CachedSelectionOrder resource on each frame.
+#[derive(Debug, Default, new)]
+pub struct CacheSelectionOrderSystem<G> {
+    phantom: PhantomData<G>,
+}
+
+impl<'a, G> System<'a> for CacheSelectionOrderSystem<G>
+where
+    G: PartialEq + Send + Sync + 'static,
+{
+    type SystemData = (
+        Entities<'a>,
+        Write<'a, CachedSelectionOrder>,
+        ReadStorage<'a, Selectable<G>>,
+    );
+    fn run(&mut self, (entities, mut cache, selectables): Self::SystemData) {
+        {
```

Comment:
Is there any way you could cut down on the work being done here by implementing some sort of dirty checking? i.e. have new selectables been added or removed, etc. I think 99% of the time this won't do any actual work

Reply:
We could eventually move to FlaggedStorage, yes.